### PR TITLE
fix($location): do not get caught in infinite digest in IE9

### LINKED ISSUE #11439: IE9 infinite $digest bug if loading a base URL without a trailing slash
I'm not honestly sure that this is reproducible in a plunker due to the specific circumstances, and unfortunately our app is secured and internal-only, so I'm not sure I can actually make a ready-to-use example.

IE9 reports infinite $digest loops in HTML5 mode when going to the base URL of an app with a default route in place if you don't use a trailing slash.  This does not happen in browsers that actually support the History API and don't need the hashbangs.

Say my base URL is http://www.mysite.com/appBase and I have an angular app hosted there with a default route set to go to /Home.  If I go to "http://www.mysite.com/appBase/" (note the trailing slash), I get redirected to http://www.mysite.com/appBase/Home and everything is perfectly all right.  No errors in the console, etc.

If I instead go to "http://www.mysite.com/appBase" (note no trailing slash), I end up in the infinite $digest loop.  Prior to angular 1.3.6, this case ended up pegging IE so that you had to force-kill (fun!).  After 1.3.6 (I've tested up through 1.3.15), we still get an infinite $digest loop but it does finally bail out and succeed, so the problem is more on the order of performance and error log flooding rather than complete disaster like it was prior to 1.3.6.  (In fact, I thought this was fixed for some time, but I must only have been testing locally and not on our deployed server with the separate base URL.)

It appears to have something to do with the $$rootScope.$watch(function $locationWatch()) call.  I *think* what's happening (and mind you, debugging anything in IE9 on a deployed server is akin to self-inflicting a root canal) is that something is losing track of the hash symbol.  I end up seeing calls to the $browser.url() setter that go back and forth between http://www.mysite.com/appBase/#/ and http://www.mysite.com/appBase/#/Home/.  Eventually *something* changes (I'm wondering if it's simply the $watch bombing due to infinite $digests) to get it out of the cycle, and things in the app are fine.

In $browser.url()'s setter, everything ends up okay if I change the following lines:

```javascript
if (!sameBase) {
    reloadLocation = url;
}
```

to:

```javascript
if (replace || !sameBase) {
    reloadLocation = url;
}
```

"replace" is getting set to true twice in a row, the first time to http://www.mysite.com/appBase/#/ and the second time to http://www.mysite.com/appBase/#/Home/.  Something is checking $browser.url()'s getter, and with that change, that place is seeing the changed reloadLocation variable after the second call with replace=true and ends up cutting out of the $digest loop.

I don't know if there are any other implications to that change, though, or if this is the right way to go about this.  Without the ability to step through in a real debugger (since IE9's debugger seems to be useless for SPAs), I'm limited to what I've been able to log to the console or pop up in an alert (since apparently console.log doesn't always fire in IE9?!?).  All I know is that Chrome and Firefox are still fine after that change as well, and I can't think of a compelling reason that we should hang on to the first "reloadLocation" if we're doing a "replace" twice.

I hope this makes at least some sense to someone.  Again, if I had the means to make a reproducible testcase, I would, but it seems to be something that's hard to emulate without free reign to muck up a public server.

### PR DESCRIPTION
**Pull request by petebacondarwin**
The problem appears to occur when we have triggered a reload (via replace()) that updates the `reloadLocation` but then before the reload happens we trigger another change to the URL that would not have triggered a reload if we had already reloaded. In this case, the `$browser` is ignoring the `reloadLocation` because it thinks that we will no longer reload.

So to be more explicit:

Navigate to http://my.server.com/Home
Because we are on IE9 this get rewritten (with replace()) to http://my.server.com/#/Home, which would cause a reload on IE9
The SuccessChangeHandler kicks in and then changes the $location.path to /, i.e. the URL gets changed to http://my.server.com/#/`.
Since http://my.server.com/#/Home -> http://my.server.com/#/ would not normally trigger a reload we don't update the reloadLocation, which gets stuck at http://my.server.com/#/Home
We get stuck in an infinite digest as the $location keeps trying to update the URL to http://my.server.com/#/ but failing because the reloadLocation is stuck at http://my.server.com/#/Home
This is why @hamfastgamgee's fix appears to work. When we see that we are in an infinite loop and there is a reloadLocation then we force it to change to the proper URL.

A better solution is simply to realize that we are updating a URL in $browser that is expected to cause a reload by changing the line at https://github.com/angular/angular.js/blob/v1.4.0/src/ng/browser.js#L150:

```
  if ($sniffer.history && (!sameBase || !sameState)) {
    history[replace ? 'replaceState' : 'pushState'](state, '', url);
    cacheState();
    // Do the assignment again so that those two variables are referentially identical.
    lastHistoryState = cachedState;
  } else {
    if (!sameBase) {
      reloadLocation = url;
    }
    if (replace) {
      location.replace(url);
    } else if (!sameBase) {
      location.href = url;
    } else {
      location.hash = getHash(url);
    }
  }
```

to

```
  if ($sniffer.history && (!sameBase || !sameState)) {
    history[replace ? 'replaceState' : 'pushState'](state, '', url);
    cacheState();
    // Do the assignment again so that those two variables are referentially identical.
    lastHistoryState = cachedState;
  } else {
    if (!sameBase || reloadLocation) {
      reloadLocation = url;
    }
    if (replace) {
      location.replace(url);
    } else if (!sameBase) {
      location.href = url;
    } else {
      location.hash = getHash(url);
    }
  }
```

In other words we check not only whether the base has changed but also whether we are in the middle of a reload any way.

Closes #11439
Closes #11935
Closes #11675

### CODE REVIEW

**Comment by petebacondarwin** [comment]
@hamfastgamgee can you take a look at this alternative?

File: test/ng/locationSpec.js [hunk 1]
```diff
@@ -663,6 +663,22 @@ describe('$location', function() {
...
+      initService({html5Mode:true,supportHistory:false});
+      mockUpBrowser({initialUrl:'http://server/base/home', baseHref:'/base/'});
+      inject(
+        function($browser, $location, $rootScope, $window) {
+          $rootScope.$on('$locationChangeSuccess', function() {
+            if ($location.path() !== '/') {
+                $location.path('/').replace();
+            }
+          });
+          $rootScope.$digest();
```
Comment (petebacondarwin) [inline test/ng/locationSpec.js]: need to check that the change handler actually gets called and that the path is updated correctly

File: test/ng/locationSpec.js [hunk 2]
```diff
@@ -672,6 +688,162 @@ describe('$location', function() {
...
+          expect($location.path()).toEqual('/');
+          expect($browserUrl.calls.length).toEqual(1);
+        });
+      });
+    });
+
+
+    describe('location watch for HTML5 browsers', function() {
+      beforeEach(initService({html5Mode: true, supportHistory: true}));
+      beforeEach(inject(initBrowser({basePath: '/app/'})));
```
Comment (petebacondarwin) [inline test/ng/locationSpec.js]: need to update these tests to use `mockupBrowser()`

File: test/ng/locationSpec.js [hunk 3]
```diff
@@ -1040,6 +1212,49 @@ describe('$location', function() {
...
+        function($browser, $location, $rootScope, $window) {
+          $rootScope.$on('$locationChangeSuccess', function() {
+            if ($location.path() !== '/') {
+                $location.path('/').replace();
+            }
+          });
+          $rootScope.$digest();
+        }
+      );
+    });
```
Comment (petebacondarwin) [inline test/ng/locationSpec.js]: Need to remove this test as it duplicates the above

**Comment by hamfastgamgee** [comment]
Well, leaving aside the failed tests, I can verify that it does in fact fix my case.  I had tried a similar fix previously (never exactly this; it was mostly removing the sameBase check) and had at least three failed tests, but it appears that your fix at least satisfies most browsers on travis. :)

**Comment by petebacondarwin** [comment]
OK, I'll fix up these tests and then if you can give it a once-over, I'll be able to merge it tomorrow

**Comment by hamfastgamgee** [comment]
Sure thing.  Can this please be backported into the 1.3.x branch as well?
My app might be able to go up to 1.4.x, but at present we're not allocating
any time to do the level of regression testing that would make our business
users comfortable.  (I realize it should be a relatively easy upgrade when
we have that time.)

**Comment by petebacondarwin** [comment]
I think this should be able to be back-ported

**Comment by petebacondarwin** [comment]
Updated.

**Comment by hamfastgamgee** [comment]
LGTM.  Still works in my app. :)

**Comment by petebacondarwin** [comment]
If Travis is green by the morning I will merge.

**Comment by IgorMinar** [comment]
lgtm especially with all the tests!

**Comment by petebacondarwin** [comment]
And it's in! Backporting to 1.3...
Then to check whether all those other infinite digest issues are fixed too...
